It turns out that the dtc update from 1.4.5 to 1.4.7 broke the build of U-Boot 2017.09, and I can't see an easy fix for it. I will probably end up adding an override to pin dtc to 1.4.5 for the Rock{Pro,}64 U-Boot builds, but for now I just added a revert for that commit in case anyone wants to test this PR.

Here is the error in case anyone has an idea of how to fix it:
```
In file included from ./tools/../lib/libfdt/libfdt.h:10:0,
                 from tools/../include/libfdt.h:1,
                 from tools/fdt_host.h:11,
                 from tools/imagetool.h:24,
                 from tools/aisimage.c:8:
/nix/store/yyfzjvpk0ij1ixb5hlaznin54sw2awpz-dtc-1.4.7/include/libfdt_env.h:71:30: error: conflicting types for 'fdt64_t'
 typedef uint64_t FDT_BITWISE fdt64_t;
                              ^~~~~~~
In file included from <command-line>:0:0:
././include/libfdt_env.h:19:16: note: previous declaration of 'fdt64_t' was here
 typedef __be64 fdt64_t;
                ^~~~~~~
In file included from ././include/libfdt_env.h:12:0,
                 from <command-line>:0:
/nix/store/yyfzjvpk0ij1ixb5hlaznin54sw2awpz-dtc-1.4.7/include/libfdt_env.h:91:24: error: expected ')' before 'x'
 static inline uint32_t fdt32_to_cpu(fdt32_t x)
                        ^
././include/compiler.h:71:9: error: expected ')' before '&' token
  ((((x) & 0xff000000) >> 24) | \
         ^
././include/compiler.h:71:23: error: expected ')' before '>>' token
  ((((x) & 0xff000000) >> 24) | \
                       ^
././include/compiler.h:71:30: error: expected ')' before '|' token
  ((((x) & 0xff000000) >> 24) | \
                              ^
/nix/store/yyfzjvpk0ij1ixb5hlaznin54sw2awpz-dtc-1.4.7/include/libfdt_env.h:95:23: error: expected ')' before 'x'
 static inline fdt32_t cpu_to_fdt32(uint32_t x)
                       ^
././include/compiler.h:71:9: error: expected ')' before '&' token
  ((((x) & 0xff000000) >> 24) | \
         ^
././include/compiler.h:71:23: error: expected ')' before '>>' token
  ((((x) & 0xff000000) >> 24) | \
                       ^
././include/compiler.h:71:30: error: expected ')' before '|' token
  ((((x) & 0xff000000) >> 24) | \
                              ^
/nix/store/yyfzjvpk0ij1ixb5hlaznin54sw2awpz-dtc-1.4.7/include/libfdt_env.h:100:24: error: expected ')' before 'x'
 static inline uint64_t fdt64_to_cpu(fdt64_t x)
                        ^
././include/compiler.h:76:9: error: expected ')' before '&' token
  ((((x) & 0xff00000000000000##sfx) >> 56) | \
         ^
././include/compiler.h:76:36: error: expected ')' before '>>' token
  ((((x) & 0xff00000000000000##sfx) >> 56) | \
                                    ^
././include/compiler.h:76:43: error: expected ')' before '|' token
  ((((x) & 0xff00000000000000##sfx) >> 56) | \
                                           ^
/nix/store/yyfzjvpk0ij1ixb5hlaznin54sw2awpz-dtc-1.4.7/include/libfdt_env.h:104:23: error: expected ')' before 'x'
 static inline fdt64_t cpu_to_fdt64(uint64_t x)
                       ^
././include/compiler.h:76:9: error: expected ')' before '&' token
  ((((x) & 0xff00000000000000##sfx) >> 56) | \
         ^
././include/compiler.h:76:36: error: expected ')' before '>>' token
  ((((x) & 0xff00000000000000##sfx) >> 56) | \
                                    ^
././include/compiler.h:76:43: error: expected ')' before '|' token
  ((((x) & 0xff00000000000000##sfx) >> 56) | \
                                           ^
make[1]: *** [scripts/Makefile.host:116: tools/aisimage.o] Error 1
make: *** [Makefile:1278: tools] Error 2
```

--- a/pkgs/misc/arm-trusted-firmware/default.nix
+++ b/pkgs/misc/arm-trusted-firmware/default.nix
@@ -1,4 +1,4 @@
-{ stdenv, fetchFromGitHub, buildPackages }:
+{ stdenv, fetchFromGitHub, pkgsCross, buildPackages }:
 
 let
   buildArmTrustedFirmware = { filesToInstall
@@ -21,6 +21,9 @@ let
     };
 
     depsBuildBuild = [ buildPackages.stdenv.cc ];
+
+    # For Cortex-M0 firmware in RK3399
+    nativeBuildInputs = [ pkgsCross.arm-embedded.stdenv.cc ];
 
     makeFlags = [
       "CROSS_COMPILE=${stdenv.cc.targetPrefix}"
@@ -80,6 +83,13 @@ in rec {
   armTrustedFirmwareRK3328 = buildArmTrustedFirmware rec {
     extraMakeFlags = [ "bl31" ];
     platform = "rk3328";
+    extraMeta.platforms = ["aarch64-linux"];
+    filesToInstall = [ "build/${platform}/release/bl31/bl31.elf"];
+  };
+
+  armTrustedFirmwareRK3399 = buildArmTrustedFirmware rec {
+    extraMakeFlags = [ "bl31" ];
+    platform = "rk3399";
     extraMeta.platforms = ["aarch64-linux"];
     filesToInstall = [ "build/${platform}/release/bl31/bl31.elf"];
   };

--- a/pkgs/misc/uboot/default.nix
+++ b/pkgs/misc/uboot/default.nix
@@ -4,7 +4,8 @@
 }:
 
 let
-  buildUBoot = { filesToInstall
+  buildUBoot = { version ? "2018.09"
+            , filesToInstall
             , installDir ? "$out"
             , defconfig
             , extraConfig ? ""
@@ -14,8 +15,8 @@ let
             , ... } @ args:
            stdenv.mkDerivation (rec {
 
-    name = "uboot-${defconfig}-${version}";
-    version = args.version or "2018.09";
+    pname = "uboot-${defconfig}";
+    inherit version;
 
     src = fetchurl {
       url = "ftp://ftp.denx.de/pub/u-boot/u-boot-${version}.tar.bz2";

--- a/pkgs/misc/uboot/default.nix
+++ b/pkgs/misc/uboot/default.nix
@@ -7,6 +7,7 @@ let
   buildUBoot = { filesToInstall
             , installDir ? "$out"
             , defconfig
+            , extraConfig ? ""
             , extraPatches ? []
             , extraMakeFlags ? []
             , extraMeta ? {}
@@ -50,10 +51,14 @@ let
       "CROSS_COMPILE=${stdenv.cc.targetPrefix}"
     ] ++ extraMakeFlags;
 
+    passAsFile = [ "extraConfig" ];
+
     configurePhase = ''
       runHook preConfigure
 
       make ${defconfig}
+
+      cat $extraConfigPath >> .config
 
       runHook postConfigure
     '';
@@ -242,10 +247,8 @@ in rec {
     extraMeta.platforms = ["armv7l-linux"];
     filesToInstall = ["u-boot-with-nand-spl.imx"];
     buildFlags = "u-boot-with-nand-spl.imx";
-    postConfigure = ''
-      cat >> .config << EOF
+    extraConfig = ''
       CONFIG_CMD_SETEXPR=y
-      EOF
     '';
     # sata init; load sata 0 $loadaddr u-boot-with-nand-spl.imx
     # sf probe; sf update $loadaddr 0 80000

--- a/pkgs/misc/uboot/rock64.nix
+++ b/pkgs/misc/uboot/rock64.nix
@@ -6,7 +6,6 @@
     sha256 = "189f7h6wj2yrcc5ga103jnyysykf9j3j9p1vcy7791bxwxqxnggf";
   };
 in buildUBoot rec {
-  name = "uboot-${defconfig}-${version}";
   version = "2017.09";
 
   src = fetchFromGitHub {

--- a/pkgs/misc/uboot/rockpro64.nix
+++ b/pkgs/misc/uboot/rockpro64.nix
@@ -6,7 +6,6 @@
     sha256 = "189f7h6wj2yrcc5ga103jnyysykf9j3j9p1vcy7791bxwxqxnggf";
   };
 in buildUBoot rec {
-  name = "uboot-${defconfig}-${version}";
   version = "2017.09";
 
   src = fetchFromGitHub {

--- a/pkgs/top-level/all-packages.nix
+++ b/pkgs/top-level/all-packages.nix
@@ -14254,6 +14254,7 @@ in
     armTrustedFirmwareAllwinner
     armTrustedFirmwareQemu
     armTrustedFirmwareRK3328
+    armTrustedFirmwareRK3399
     ;
 
   microcodeAmd = callPackage ../os-specific/linux/microcode/amd.nix { };

